src/gradualizer_lib.erl

```diff
@@ -83,58 +84,86 @@ reverse_graph(G) ->
     from_edges(maps:keys(G), [ {J, I} || {I, Js} <- maps:to_list(G), J <- Js ]).
 
 
+get_user_type_definition(Types, Anno, Name, Args) ->
+    %% Let's check if the type is a known remote type.
+    case typelib:get_module_from_annotation(Anno) of
+        {ok, Module} ->
+            case gradualizer_db:get_opaque_type(Module, Name, Args) of
+                {ok, Ty} ->
+                    {ok, Ty};
+                not_found ->
+                    not_found
+            end;
+        none ->
+            %% Let's check if the type is defined in the context of this module.
+            case maps:get({Name, length(Args)}, Types, not_found) of
+                {_Params, Ty} ->
+                    {ok, Ty};
```

Comment:
Don't we need to substitute Args for Params in Ty here?

Reply:
Hmmm, I'm not sure yet, I'd have to devise a test to work it out on some examples...

Reply:
Yes, tests are good to have.

Just thinking that `gradualizer_db:get_type/3` does substitute type params and that local and remote types probably should be handled in the same way.

Reply:
Btw, this clause in `typechecker` is very similar and can maybe be refactored to use `get_user_type_definition`:

```Erlang
normalize({user_type, P, Name, Args} = Type, TEnv) ->
    case typelib:get_module_from_annotation(P) of
        {ok, Module} ->
            %% Local type in another module, from an expanded remote type
            case gradualizer_db:get_type(Module, Name, Args) of
                {ok, T} ->
                    normalize(typelib:remove_pos(T), TEnv);
                opaque ->
                    Type;
                not_found ->
                    throw({undef, user_type, P, {Module, Name, length(Args)}})
            end;
        none ->
            %% Local user-defined type
            TypeId = {Name, length(Args)},
            case TEnv#tenv.types of
                #{TypeId := {Vars, Type0}} ->
                    VarMap = maps:from_list(lists:zip(Vars, Args)),
                    Type1 = typelib:substitute_type_vars(Type0, VarMap),
                    normalize(Type1, TEnv);
                _NotFound ->
                    throw({undef, user_type, P, {Name, length(Args)}})
            end
    end;
```

Reply:
Ok, this one is refactored to use `get_type_definition`, too.